### PR TITLE
Reuse playback buffers in FluidSynth

### PR DESCRIPTION
**Pull request by kcgen**
Overall PR is quite small and reviewable commit-by-commit. 

I've implemented this in FluidSynth right now; plan is to:
1. review this implementation first,
1. address feedback and refine as needed, and
1. replicate the finalized approach for MT-32 as a final commit in this PR.

### CODE REVIEW

File: src/midi/midi_fluidsynth.cpp [hunk 1]
```diff
@@ -426,14 +426,19 @@ uint16_t MidiHandlerFluidsynth::GetRemainingFrames()
...
+	std::vector<int16_t> playable_buffer(SAMPLES_PER_BUFFER);
+
 	while (keep_rendering) {
 		fluid_synth_write_float(synth.get(), FRAMES_PER_BUFFER,
 		                        render_buffer.data(), 0, 2, render_buffer.data(), 1, 2);
-		auto out = soft_limiter.Process(render_buffer, FRAMES_PER_BUFFER);
-		ring.wait_enqueue(out);
+		soft_limiter.Process(render_buffer, FRAMES_PER_BUFFER,
+		                     playable_buffer);
+		ring.wait_enqueue(playable_buffer);
```
Comment (dreamer) [inline src/midi/midi_fluidsynth.cpp]: I might've been wrong in suggesting to avoid creating temporary objects here…

Which `wait_enqueue` is being called in here?

`void wait_enqueue(T const& item)` (in which case, according to rwqueue documentation, the vector is copied, so we get limited benefit from reusing the object. <- I hope this is the case. (It's not necessarily bad that it's copied - copying larger vector like this probably happens via fast memcpy copying whole memory blocks at once).

or `void wait_enqueue(T&& item)` (in which case our locally-created `playable_buffer` is moved out, so starting with 2nd iteration we're writing over random memory while filling it in inside `Process`)?

*Side note: I am starting to think we don't need atomic rwqueue at all - we need much simpler structure implementing double or tripple-buffering. Perhaps it already works like that, just with unnecessary steps along the way.*

File: src/midi/midi_mt32.cpp [hunk 2]
```diff
@@ -457,7 +457,7 @@ void MidiHandler_mt32::Render()
 	while (keep_rendering) {
 		service->renderFloat(render_buffer.data(), FRAMES_PER_BUFFER);
 		soft_limiter.Process(render_buffer, FRAMES_PER_BUFFER, playable_buffer);
-		ring.wait_enqueue(playable_buffer);
+		ring.wait_enqueue(std::move(playable_buffer));
```
Comment (dreamer) [inline src/midi/midi_mt32.cpp:460]: OK, now it's definitely wrong - we can't use `playable_buffer` after moving it out! We would need to re-initialize the vector!

File: src/midi/midi_fluidsynth.cpp (hunk 1 above)
Comment (kcgen) [reply src/midi/midi_fluidsynth.cpp]: > I might've been wrong in suggesting to avoid creating temporary objects here…

The PR is a three-step transition; each one is compilable. In this first commit, we're just adjusting soft limiter (in the commit comment). 

> Which wait_enqueue is being called in here?

Correct, this is still the copy call being used in the first commit; I just changing one thing at a time, and keeping the code buildable and testable at each step.

File: src/midi/midi_mt32.cpp (hunk 2 above)
Comment (kcgen) [reply src/midi/midi_mt32.cpp:460]: > can't use playable_buffer without re-initializing

Exactly :-)  in this second commit we want prove that move into rw-queue is occurring, so inside the soft limiter, it will re-initialize this inbound play buffer and print a log message 

``` c++
(In soft limiter process function)
if (out.size() < frames) {
	out.clear(); // re-initialize
	out.resize(frames);
	DEBUG_LOG_MSG("SOFT_LIMITER: reinitialized output buffer");
}
```
In the commit, I mention this this is an interim step, and of course we wouldn't want code like this.. but it's just a proof point. 

You'll see in the 3rd commit this become an assert just to check the size, because in the 3rd commit we achieve perfect move rotation of the 8 buffers in and out of the two rw-queues.

File: src/midi/midi_fluidsynth.cpp (hunk 1 above)
Comment (kcgen) [reply src/midi/midi_fluidsynth.cpp]: > Side note: I am starting to think we don't need atomic rwqueue at all - we need much simpler structure implementing double or tripple-buffering. Perhaps it already works like that, just with unnecessary steps along the way.

The rw-queue's ability to let us move objects in and out of a simple queue in two places of code (that are actually being run in separate threads) belies the work it's doing under the hood to give us such an easier to understand workflow up in our code.

  It's easy to forget it's even multithreaded (I'm not saying you are), and the code makes it easy to envision using different container or queue in it's place, which is testament to how easy it's made the code to read. 

By the 3rd commit, you will see the batch of 8 allocated buffers being rotated in and out of two queues, all using moves.  It's a sweet as it gets :-) this would normally be significantly harder without the rw-queue. 

But as for experimenting with other structures or buffering techques - please go for it! By the 3rd commitw, it's quite readable exactly how the buffers are rotating so should be easy to try new ideas :-) 👍

**Comment by kcgen** [comment]
Still need to benchmark against master.
Plan is to add timers in both the callback and rendering threads.

**Comment by kcgen** [comment]
# Benchmarks

Same callback-latency benchmark procedure here - https://github.com/dosbox-staging/dosbox-staging/pull/887

| Branch | Test | median | avg | stdev (ms) | max (ms) | log |
| --- | --- | --- | --- | --- | --- | --- |
| master | idle | 2.12  |   3.89  |  26.11 | 2808.81 | [master-idle.txt](https://github.com/dosbox-staging/dosbox-staging/files/6017998/master-idle.txt) |
| PR | idle | 1.87   |  3.57  |  10.82 | 1665.97 | [pr-idle.txt](https://github.com/dosbox-staging/dosbox-staging/files/6017999/pr-idle.txt) |
| master | load | 3.10   |  7.59  |  142.09 | 17578.48 | [master-load.txt](https://github.com/dosbox-staging/dosbox-staging/files/6018000/master-load.txt) |
| PR | load | 3.04  |  6.05  | 73.50  | 6789.41 | [pr-load.txt](https://github.com/dosbox-staging/dosbox-staging/files/6018001/pr-load.txt) |

Although both branches provide low average latency, the standard deviation and maximum latency has dropped in the PR branch.  This indicates that adding dynamic allocation and copy-behaviour introduces some unpredictable amount of extra work; but only sporadically (perhaps when cache-line stalls occur).

**Comment by dreamer** [comment]
@kcgen It all makes sense now :) Great change!

I also re-tested on HoMM2, as that's the game that was notoriously having weird edge-case issues with MIDI - as far as I can tell, it's perfect now. Tested on a few other titles and it all went smooth as well.

**Review by dreamer — APPROVED** [review]

**Comment by kcgen** [comment]
Thanks @dreamer!   

I tested our original 0.76 release versus this, as this is the end of the line performance-wise (before sacrificing code quality).

I also tested under the Linux's lower-latency scheduling profiles to see if it could further improve the numbers.
 - **Default scheduling**: DOSBox is equal runtime rights as other user processes (on idle system)
 - **Batch scheduling**: DOSBox given increased time-slice allowance to perform more work over other processes
 - **Realtime FIFO**:  DOSBox is allowed remain running ahead of all other processes (even kernel processes) except for IO interrupts and other higher-priority FIFO processes. 
 -  **Realtime Round-Robin**: DOSBox similar to FIFO, but hard control on latency is added to each thread (each running for a fixed 'quantum' amount of time before cycling back again).

All times in micro-seconds except for max, in milliseconds.

| Version | Scheduling | median | avg | stdev |  max (ms) |
| --- | --- | ---: | ---: | ---: | ---: | 
| 0.76 | Default |  131.96 |  245.78 | 1106.50 | 12.46 |
| 0.76 | Batch |  109.87  | 175.95  | 309.60 | 9.62 |
| 0.76 | RT-FIFO | 125.80  | 144.46 |  121.30  | 0.87 |
| 0.76 | RT-RR |   111.98 |  161.58  | 149.44  | 0.70 |
| 0.77α | Default |  1.95 |    3.30  |  13.11 |  2.07 |
| 0.77α | Batch |  1.92  |  3.17 | 7.22 | 0.92 |
| 0.77α | RT-FIFO |  1.89 |  2.69 | 1.85 | 0.08 | 
| 0.77α | RT-RR |   1.95 |  2.71 | 1.75 | 0.08 |
